canonical/cloud-init
# Allow comments in runcmd and report failed commands correctly.

## Proposed Commit Message


```
summary: Allow comments in runcmd and report failed commands correctly.

A `runcmd` script may fail to parse properly, but does not mark `runcmd` as failed when that occurs.
Additionally `shellify()` fails to correctly parse scripts that contain a comment line.

Rectify both issues and add unit tests to verify correct behavior.

LP: #1853146
```

## Test Steps
To reproduce these issues, use the following `runcmd` and check the logs after cloud-init runs.
```
runcmd:
   - echo start
   - # 
   - echo end
```

This PR contains four commits. The tests added in the first and third test will fail unless the second and fourth fix commits are applied. Rolling back to these commits from HEAD (`git reset HEAD~3` && `git reset HEAD~1`) before running the new tests should demonstrate the failures. The tests should succeed from the second and fourth commit in this series.

## Checklist:

 - [ TODO: Sign the CLA ] My code follows the process laid out in [the documentation](https://cloudinit.readthedocs.io/en/latest/topics/hacking.html)
 - [ x ] I have updated or added any unit tests accordingly
 - [ x ] I have updated or added any documentation accordingly
